TOC Restructuring Final PR (#2402)
couchbase/docs-server: modules/fts/pages/fts-search-request.adoc

[#Consistency]
== Consistency

[abstract]
A mechanism to ensure that the Full Text Search (FTS) index can obtain the most up-to-date version of the document written to a collection or a bucket. 

The consistency mechanism provides xref:#consistency-vectors[Consistency Vectors] as objects in the search query that ensures FTS index searches all your last data written to the vBucket. 

The search service does not respond to the query until the designated vBucket receives the correct sequence number. 

The search query remains blocked while continuously polling the vBucket for the requested data. Once the sequence number of the data is obtained, the query is executed over the data written to the vBucket.

When using this consistency mode, the query service will ensure that the indexes are synchronized with the data service before querying.

=== Workflow to understand Consistency

. Create an FTS index in Couchbase.
. Write a document to the Couchbase cluster. 
. Couchbase returns the associate vector to the app, which needs to issue a query request with the vector.
. The FTS index starts searching the data written to the vBucket.

> In this workflow, it is possible that the document written to vBucket is not yet indexed. So, when FTS starts searching that document, the most up-to-date document versions are not retrieved, and only the indexed versions are queried.

Reviewer: ```suggestion
In this workflow, it is possible that the document written to the vBucket is not yet indexed. So, when FTS starts searching that document, the most up-to-date document versions are not retrieved, and only the indexed versions are queried.
```

Author: updated

> Therefore, the Couchbase server provides a consistency mechanism to overcome this issue and ensures that the FTS index can search the most up-to-date document written to vBucket.

Reviewer: ```suggestion
Therefore, the Couchbase server provides a consistency mechanism to overcome this issue and ensures that the FTS index can search the most up-to-date document written to the vBucket.
```

Author: updated

=== Consistency Level

The consistency level is a parameter that either takes an empty string indicating the unbounded (not_bounded) consistency or at_plus indicating the bounded consistency.

==== at_plus

Executes the query, requiring indexes first to be updated to the timestamp of the last update. 

This implements bounded consistency. The request includes a scan_vector parameter and value, which is used as a lower bound. This can be used to implement read-your-own-writes (RYOW).

If index-maintenance is running behind, the query waits for it to catch up.

==== not_bounded

Executes the query immediately, without requiring any consistency for the query. No timestamp vector is used in the index scan. 

This is the fastest mode, because it avoids the costs of obtaining the vector and waiting for the index to catch up to the vector.

If index-maintenance is running behind, out-of-date results may be returned.

[#consistency-vectors]
=== Consistency Vectors

> The consistency vectors supporting the consistency mechanism in Couchbase contain the mapping of the vbucket and sequence number of the data stored in the vBucket.

Reviewer: ```suggestion
The consistency vectors supporting the consistency mechanism in Couchbase contain the mapping of the vBucket and sequence number of the data stored in the vBucket.
```

[source, JSON]
----
{
  "ctl": {
    "timeout": 10000,
    "consistency": {
      "vectors": {
        "index1": {
          "607/205096593892159": 2,
          "640/298739127912798": 4
        }
      },
      "level": "at_plus"
    }
  },
  "query": {
    "match": "jack",
    "field": "name"
  }
}
----

=== Consistency Results

Consistency result is the attribute that you can use to set the query result option, such as complete.

==== Example:
[source, JSON]
----
{
  "query": {...}, 
  "ctl": {
    "consistency": {
      "results": "complete"
    }
  }
} 
----

=== The "Complete" option

The complete option allows you to set the query result as "complete" which indicates that if any of the index partitions are unavailable due to the node not being reachable, the query will display an error in response instead of partial results.
    
==== Example
[source, JSON]
----
{
  "query": {...}, 
  "ctl": {
    "consistency": {
      "results": "complete"
    }
  }
}
----

=== Consistency Tips and Recommendations

> Consistency vectors provide 'read your own writes' functionality where the read operation waits for a specific time until the write operation is finished.

Reviewer: ```suggestion
Consistency vectors provide "read your own writes" functionality where the read operation waits for a specific time until the write operation is finished.
```

Reviewer: I think the change above this one used double-quotes, so double quotes should be used throughout.

Author: updated

There are multiple ways to achieve pagination with settings within a search request. Pagination will fetch a deterministic set of results when the results are sorted in a certain fashion.

Pagination provides the following options: 

=== Size/from or offset/limit

This pagination settings can be used to obtain a subset of results and works deterministically when combined with a certain sort order.

Using `size/limit` and `offset/from` would fetch at least `size + from` ordered results from a partition and then return the `size` number of results starting at offset `from`.

Deep pagination can therefore get pretty expensive when using `size + from` on a sharded index due to each shard having to possibly return large resultsets (at least `size + from`) over the network for merging at the coordinating node before returning the `size` number of results starting at offset `from`.

The default sort order is based on _score_ (relevance) where the results are ordered from the highest to the lowest score.

==== Example

Here's an example query that fetches results from the 11th onwards to the 15th that have been ordered by _score_.

> ----

Reviewer: Add [source, json]

Author: added

{
  "query": {
      "match": "California",
      "field": "state"
  },
  "size": 5,
  "from": 10
}
----

== Search after/before

For an efficient pagination, you can use the `search_after/search_before` settings.

`search_after` is designed to fetch the `size` number of results after the key specified and `search_before` is designed to fetch the `size` number of results before the key specified.

These settings allow for the client to maintain state while paginating - the sort key of the last result (for search_after) or the first result (for search_before) in the current page.

Both the attributes accept an array of strings (sort keys) - the length of this array will need to be the same length of the "sort" array within the search request.

NOTE: You cannot use both `search_after` and `search_before` in the same search request.

=== Example

Here are some examples using `search_after/search_before` over sort key "_id" (an internal field that carries the document ID).

> ----

Reviewer: [source, json]

Author: added

{
  "query": {
      "match": "California",
      "field": "state"
  },
  "sort": ["_id"],
  "search_after": ["hotel_10180"],
  "size": 3
}
----

> ----

Reviewer: [source, json]

Author: added

{
  "query": {
      "match": "California",
      "field": "state"
  },
  "sort": ["_id"],
  "search_before": ["hotel_17595"],
  "size": 4
}
----

NOTE: A Full Text Search request that doesn't carry any pagination settings will return the first 10 results (`"size: 10", "from": 0`) ordered by _score_ sequentially from the highest to lowest.

=== Pagination tips and recommendations

> The pagination of search results can be done using the 'from' and 'size' parameters in the search request. But as the search gets into deeper pages, it starts consuming more resources.  

Reviewer: ```suggestion
The pagination of search results can be done using the `from` and `size` parameters in the search request. But as the search gets into deeper pages, it starts consuming more resources.  
```

Author: updated

To safeguard against any arbitrary higher memory requirements, FTS provides a configurable limit bleveMaxResultWindow (10000 default) on the maximum allowable page offsets. However, bumping this limit to higher levels is not a scalable solution.

To circumvent this problem, the concept of key set pagination in FTS, is introduced. 

> Instead of providing _from_ as a number of search results to skip, the user will provide the sort value of a previously seen search result (usually, the last result shown on the current page).  The idea is that to show the next page of the results, we just want the top N results of that sort after the last result from the previous page.

Reviewer: Consistency needed in the way variables are shown in the text. I would stick to using backticks rather than italics or quotation marks.

Author: updated the variable

This solution requires a few preconditions be met:

* The search request must specify a sort order.

NOTE: The sort order must impose a total order on the results.  Without this, any results which share the same sort value might be left out when handling the page navigation boundaries.  

A common solution to this is to always include the document ID as the final sort criteria.                                       

For example, if you want to sort by [“name”, “-age”], instead of sort by [“name”, “-age”, "_id"].

With `search_after`/`search_before` paginations, the heap memory requirement of deeper page searches is made proportional to the requested page size alone. So it reduces the heap memory requirement of deeper page searches significantly down from the offset+from values.

> _._

Reviewer: Sorry, I might have asked before, but what is this _._ for?

Author: to break the formatting

[#Sorting]
== Sorting

[abstract]
The FTS results are returned as objects. FTS query includes options to order the results.

=== Sorting Result Data

> FTS sorting is sorted by descending order of relevance.  It can , however, be customized to sort by different fields, depending on the application. 

Reviewer: ```suggestion
FTS sorting is sorted by descending order of relevance. It can, however, be customized to sort by different fields, depending on the application. 
```

These can be of three types:

* _field name_: Specifies the name of a field.
+
If multiple fields are included in the array, the sorting of documents begins according to their values for the field whose name is first in the array.
+
If any number of these values are identical, their documents are sorted again, this time according to their values for the field whose name is second; then, if any number of these values are identical, their documents are sorted a third time, this time according to their values for the field whose name is third; and so on.
+
Any document-field may be specified to hold the value on which sorting is to be based, provided that the field has been indexed in some way, whether dynamically or specifically.
+
The default sort-order is _ascending_.
If a field-name is prefixed with the `-` character, that field's results are sorted in _descending_ order.

* `_id`:Refers to the document identifier.
Whenever encountered in the array, causes sorting to occur by document identifer.

* `_score`: Refers to the score assigned the document in the result-set.
Whenever encountered in the array, causes sorting to occur by score.

==== Example

> // #Need full example here#

Reviewer: ```suggestion
```

Author: updated

* `mode`: Specifies the search-order for index-fields that contain multiple values (in consequence of arrays or multi-token analyzer-output).
The `default` order is undefined but deterministic, allowing the paging of results from `from (_offset_)`, with reliable ordering.
To sort using the minimum or maximum value, the value of `mode` should be set to either `min` or `max`.

* `type`: Specifies the type of the search-order field value. 
For example, `string` for text fields, `date` for DateTime fields, or `number` for numeric/geo fields.

To fetch more accurate sort results, we strongly recommend specifying the `type` of the sort fields in the sort section of the search request.

==== Example

The example below shows how to specify the object-sort.

[source, json]
----
{
  "explain": false,
  "fields": [
     "*"
   ],
   "highlight": {},
   "query": {
     "match": "bathrobes",
     "field": "reviews.content",
     "analyzer": "standard"
   },
   "size" : 10,
   "sort": [
      {
       "by" : "field",
       "field" : "reviews.ratings.Overall",
       "mode" : "max",
       "missing" : "last",
       "type": "number"
      }
   ]
}
----

> NOTE: The above sample assumes that the `travel-sample` bucket has been loaded, and a default index has been created on it. 

Reviewer: Should probably tell them this before you show them the sample.

Author: updated

* Date Range Facet - computes facet on the ‘updated’ field that has 2 values old and new.
+
----
curl -XPOST -H "Content-Type: application/json" -uAdministrator:asdasd http://<node>:8094/api/index/bix/query -d '{
"ctl": {"timeout": 0},
"from": 0,
"size": 0,
"query": {
            "field": "country",
            "term": "united"
},
        "facets": {
        "types": {
        "size": 10,
        "field": "updated",
        "date_ranges": [
        {
        "name": "old",
        "end": "2010-08-01"
        },
        {
        "name": "new",
        "start": "2010-08-01"
        }
]
}
}
}'
----
+
Results
+
[source,json]
----
 "facets": {
             "types": {
                "field": "updated",
                "total": 954,
                "missing": 0,
                "other": 0,
             "date_ranges": [
              {
                "name": "old",
                "end": "2010-08-01T00:00:00Z",
                "count": 934
              },
              {
                "name": "new",
                "start": "2010-08-01T00:00:00Z",
                "count": 20
              }
               ]
             }
           }
----

[#Collections]
== Collections

> Collections field let’s the user specify an optional list of collection names. This would help the users scope their search request to only those specified collections within a multi-collection index. This becomes useful with multi-collection indexes as it can speed up searches, as well as the user can manage the Role Based Access Control more granularly with this option. Ie the search user only needs permissions for the requested collections and not for every other collection indexed within the index.

Reviewer: ```suggestion
Collections field lets the user specify an optional list of collection names. This would help the users scope their search request to only those specified collections within a multi-collection index. This becomes useful with multi-collection indexes as it can speed up searches, as well as the user can manage the Role Based Access Control more granularly with this option. Ie the search user only needs permissions for the requested collections and not for every other collection indexed within the index.
```

Reviewer: Try to keep it one sentence per line; it makes it much easier to review changes.

Author: updated
